### PR TITLE
[LAA] Support different strides & non constant dep distances using SCEV.

### LINKED ISSUE #87336: [Loop Vectorizer] Loop not vectorized when accessing the same pointer with different offsets.
See reproducer in https://godbolt.org/z/v7fqorb1h.
`a2 = a1 + 256;`
`a2` has an interleaved store access, and `a1 `has a simple linear access.
There are no actual dependencies between them when VF is 4, and it looks safe for the loop to be vectorized.
BTW, changing `a2[2*i]` to `a2[i]`  makes vectorization safe.

### PR DESCRIPTION
**Pull request by fhahn**
Extend LoopAccessAnalysis to support different strides and as a consequence non-constant distances between dependences using SCEV to reason about the direction of the dependence.

In multiple places, logic to rule out dependences using the stride has been updated to only be used if StrideA == StrideB, i.e. there's a common stride.

We now also may bail out at multiple places where we may have to set FoundNonConstantDistanceDependence. This is done when we need to bail out and the distance is not constant to preserve original behavior.

Fixes https://github.com/llvm/llvm-project/issues/87336

### CODE REVIEW

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 1]
```diff
@@ -1923,8 +1923,9 @@ isLoopVariantIndirectAddress(ArrayRef<const Value *> UnderlyingObjects,
 // of various temporary variables, like A/BPtr, StrideA/BPtr and others.
 // Returns either the dependence result, if it could already be determined, or a
 // tuple with (Distance, Stride, TypeSize, AIsWrite, BIsWrite).
-static std::variant<MemoryDepChecker::Dependence::DepType,
-                    std::tuple<const SCEV *, uint64_t, uint64_t, bool, bool>>
+static std::variant<
+    MemoryDepChecker::Dependence::DepType,
+    std::tuple<const SCEV *, uint64_t, uint64_t, uint64_t, bool, bool>>
```
Comment (nikic) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:1928]: Comment above needs updating -- though I feel like at this point this tuple really wants to be a struct.

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 2]
```diff
@@ -2011,68 +2012,109 @@ MemoryDepChecker::Dependence::DepType MemoryDepChecker::isDependent(
...
+        areStridedAccessesIndependent(std::abs(Distance), CommonStride,
+                                      TypeByteSize)) {
+      LLVM_DEBUG(dbgs() << "LAA: Strided accesses are independent\n");
+      return Dependence::NoDep;
+    }
+  }
+
+  // Write to the same location with the same size.
+  if (SE.isKnownPredicate(CmpInst::ICMP_EQ, Dist,
+                          SE.getZero(Dist->getType()))) {
```
Comment (nikic) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:2052]: Don't use isKnownPredicate with ICMP_EQ -- directly check whether Dist is zero. (As such, should be part of the `if (C)` branch).

**Review by nikic — COMMENTED** [review]
(just some drive-by comments)

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp]: Done in cac4c14ecfcb892201ae8f4b1559909cc22082d0 and updated this PR, thanks!

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 2 above)
Comment (fhahn) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp]: Done, thanks!

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 3]
```diff
@@ -2028,68 +2029,105 @@ MemoryDepChecker::Dependence::DepType MemoryDepChecker::isDependent(
   if (std::holds_alternative<Dependence::DepType>(Res))
     return std::get<Dependence::DepType>(Res);
 
-  const auto &[Dist, Stride, TypeByteSize, AIsWrite, BIsWrite] =
+  const auto &[Dist, StrideA, StrideB, TypeByteSize, AIsWrite, BIsWrite] =
       std::get<DepDistanceStrideAndSizeInfo>(Res);
   bool HasSameSize = TypeByteSize > 0;
 
+  uint64_t CommonStride = StrideA == StrideB ? StrideA : 0;
```
Comment (Meinersbur) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:2036]: Isn't `0` a valid stride (i.e. indexed by a constant)? 
```suggestion
  std::optional<uint64_t> CommonStride = StrideA == StrideB ? StrideA : None;
```

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 4]
```diff
@@ -2028,68 +2029,105 @@ MemoryDepChecker::Dependence::DepType MemoryDepChecker::isDependent(
...
+    return Dependence::Unknown;
+  }
+
   ScalarEvolution &SE = *PSE.getSE();
   auto &DL = InnermostLoop->getHeader()->getModule()->getDataLayout();
-  if (!isa<SCEVCouldNotCompute>(Dist) && HasSameSize &&
+  if (HasSameSize && CommonStride &&
       isSafeDependenceDistance(DL, SE, *(PSE.getBackedgeTakenCount()), *Dist,
-                               Stride, TypeByteSize))
+                               CommonStride, TypeByteSize))
```
Comment (Meinersbur) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:2047]: `isSafeDependenceDistance` doesn't depend on there being a common stride. It could either be the larger one, or a case-distinction of either Dist positive (Src < Sink) or positive (Sink < Src).

File: llvm/lib/Transforms/Scalar/LoopLoadElimination.cpp [hunk 5]
```diff
@@ -126,8 +126,10 @@ struct StoreToLoadForwardingCandidate {
 
     // We don't need to check non-wrapping here because forward/backward
     // dependence wouldn't be valid if these weren't monotonic accesses.
-    auto *Dist = cast<SCEVConstant>(
+    auto *Dist = dyn_cast<SCEVConstant>(
         PSE.getSE()->getMinusSCEV(StorePtrSCEV, LoadPtrSCEV));
+    if (!Dist)
+      return false;
```
Comment (Meinersbur) [inline llvm/lib/Transforms/Scalar/LoopLoadElimination.cpp:132]: Is this related? LAA doesn't seem to be involved.

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 6]
```diff
@@ -2028,68 +2029,105 @@ MemoryDepChecker::Dependence::DepType MemoryDepChecker::isDependent(
...
     // couldPreventStoreLoadForward, even if it changed MinDepDistBytes,
     // since a forward dependency will allow vectorization using any width.
-    if (IsTrueDataDependence && EnableForwardingConflictDetection &&
-        (!HasSameSize || couldPreventStoreLoadForward(Val.abs().getZExtValue(),
-                                                      TypeByteSize))) {
-      LLVM_DEBUG(dbgs() << "LAA: Forward but may prevent st->ld forwarding\n");
-      return Dependence::ForwardButPreventsForwarding;
+    if (IsTrueDataDependence && EnableForwardingConflictDetection) {
+      if (!C) {
+        FoundNonConstantDistanceDependence |= CommonStride;
```
Comment (Meinersbur) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:2089]: If there is no common stride (`CommonStride == 0`) then any of the two could still be non-constant?

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 7]
```diff
@@ -2028,68 +2029,105 @@ MemoryDepChecker::Dependence::DepType MemoryDepChecker::isDependent(
...
+  if (!C) {
+    LLVM_DEBUG(dbgs() << "LAA: Dependence because of non-constant distance\n");
+    FoundNonConstantDistanceDependence |= CommonStride;
+    return Dependence::Unknown;
+  }
+
+  // The logic below currently only supports StrideA ==  StrideB, i.e. there's a
+  // common stride.
+  if (!CommonStride)
+    return Dependence::Unknown;
```
Comment (Meinersbur) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:2126]: Many checks above also test (`CommonStride > 1`). Move them below this check?

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 8]
```diff
@@ -2028,68 +2029,105 @@ MemoryDepChecker::Dependence::DepType MemoryDepChecker::isDependent(
...
   }
 
   // Negative distances are not plausible dependencies.
-  if (Val.isNegative()) {
+  if (SE.isKnownNonPositive(Dist)) {
+    if (!SE.isKnownNegative(Dist) && !HasSameSize) {
+      LLVM_DEBUG(dbgs() << "LAA: possibly zero dependence difference but "
+                           "different type sizes\n");
+      return Dependence::Unknown;
+    }
```
Comment (Meinersbur) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:2081]: My impression is that most of these `return Unknown;` should fall through instead to give other checks a chance to match. If there is an invariant after a certain point (e.g. "type size is equal") is should be rejected there. It avoids having to maintain several conditions e.g. updating `FoundNonConstantDistanceDependence` which seems to be missing here as well.

**Review by Meinersbur — COMMENTED** [review]
I would enjoy more textual description of what every condition is meant to check.

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 8 above)
Comment (fhahn) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp]: I've merged the code from above that handles C == 0 with the code here. The main reason we need to update `FoundNonConstantDistanceDependence` in multiple places now is that there now are multiple exit paths that return Unknown with non-constant dependences. 

The latest version of the patch limits setting `FoundNonConstantDistanceDependence` to 3 times, with the one before `couldPreventStoreLoadForward` potentially being removed as a follow-up by making `couldPreventStoreLoadForward` work with non-constant distances.

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 4 above)
Comment (fhahn) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp]: At the moment the comment in the function implies that it assumes equal strides which is why I left it as-is for now, to limit the scope of the initial patch. 

It tries to prove `|Dist| > BackedgeTakenCount * Stride`, so using the maximum should conservatively be correct. 

A similar reasoning can be applied at the end of the function too, which also is currently guarded by having a common stride. I am also planning on addressing that separately.

**Review by fhahn — COMMENTED** [review]
> I would enjoy more textual description of what every condition is meant to check.

There are multiple places that hand off reasoning to called functions, would you like to have a summary of what the function checks there?  Could do as separate patch, as this would be independent of the current patch?

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 4 above)
Comment (Meinersbur) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp]: Could you add a TODO?

**Comment by Meinersbur** [comment]
> > I would enjoy more textual description of what every condition is meant to check.
> 
> There are multiple places that hand off reasoning to called functions, would you like to have a summary of what the function checks there? Could do as separate patch, as this would be independent of the current patch?

I was not familiar with this code, trying to reduce the impact of this patch doesn't help me to understand it and convince myself that it does not cause miscompilation, it rather makes it even more difficult since conditions are now all over the place.

For instance, the comment
```
// Attempt to prove strided accesses independent.
```
is useless. The entire function tries to prove access independency with a possible stride. Documentation for `areStridedAccessesIndependent` (itself not descriptive) is far away. A better comment could be
```
// If the offset and strides are known constants, check whether the accesses interlace each other.
```

**Comment by fhahn** [comment]
> > > I would enjoy more textual description of what every condition is meant to check.
> > 
> > 
> > There are multiple places that hand off reasoning to called functions, would you like to have a summary of what the function checks there? Could do as separate patch, as this would be independent of the current patch?
> 
> I was not familiar with this code, trying to reduce the impact of this patch doesn't help me to understand it and convince myself that it does not cause miscompilation, it rather makes it even more difficult since conditions are now all over the place.
> 

Thanks, I put up https://github.com/llvm/llvm-project/pull/89381 to add extra documentation and updated this PR to be based on #89381

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 9]
```diff
@@ -2028,41 +2029,61 @@ MemoryDepChecker::Dependence::DepType MemoryDepChecker::isDependent(
...
+                                      TypeByteSize)) {
+      LLVM_DEBUG(dbgs() << "LAA: Strided accesses are independent\n");
+      return Dependence::NoDep;
+    }
   }
 
   // Negative distances are not plausible dependencies.
-  if (Val.isNegative()) {
+  if (SE.isKnownNonPositive(Dist)) {
+    if (!SE.isKnownNegative(Dist)) {
```
Comment (Meinersbur) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:2076]: [serious] Not knowing something doesn't imply anything.
```suggestion
    // KnownNonPositive && KnownNonNegative => KnownZero
    if (SE.isKnownNonNegative(Dist)) {
```

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 10]
```diff
@@ -2071,34 +2092,51 @@ MemoryDepChecker::Dependence::DepType MemoryDepChecker::isDependent(
...
-  if (Val == 0) {
-    if (HasSameSize)
-      return Dependence::Forward;
-    LLVM_DEBUG(
-        dbgs() << "LAA: Zero dependence difference but different type sizes\n");
+  if (!C) {
+    // TODO: Relax requirement that there is a common stride to retry with
+    // non-constant distance dependencies.
+    FoundNonConstantDistanceDependence |= !!CommonStride;
+    LLVM_DEBUG(dbgs() << "LAA: Dependence because of non-constant distance\n");
```
Comment (Meinersbur) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:2119]: This is the bail-out moved here from line 2047. Is this done so return Dependence::Forward with negative distance can be returned also on non-constant distances?

`FoundNonConstantDistanceDependence` is not set if there is no common stride, but what does it have to do with the distance between the pointers? Before this patch it would set that flag independent of any stride.

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 9 above)
Comment (fhahn) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp]: Thanks, updated!

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 11]
```diff
@@ -2028,41 +2029,61 @@ MemoryDepChecker::Dependence::DepType MemoryDepChecker::isDependent(
...
+  }
+
   ScalarEvolution &SE = *PSE.getSE();
   auto &DL = InnermostLoop->getHeader()->getModule()->getDataLayout();
   // If the distance between the acecsses is larger than their absolute stride
   // multiplied by the backedge taken count, the accesses are independet, i.e.
   // they are far enough appart that accesses won't access the same location
   // across all loop ierations.
-  if (!isa<SCEVCouldNotCompute>(Dist) && HasSameSize &&
+  if (HasSameSize && CommonStride &&
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp:2054]: Yep, I plan to address this separately.

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 10 above)
Comment (fhahn) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp:2123]: It is independent of whether there is a common stride or not, but previously `getDependenceDistanceStrideAndSize`  would exit early with `Unknown` if the strides weren't matching, so `FoundNonConstantDistanceDependence` was only ever set if there was a common stride which is what the current patch tries to preserve in the initial version.

The current `FoundNonConstantDistanceDependence` can lead to runtime check generation in cases where the runtime checks will always fail unfortunately, which is why I'd like to keep its original behavior for now, as we regress codegen otherwise in some cases.

**Review by fhahn — COMMENTED** [review]
The latest comments should be addressed/responded to, thanks!

I also added a missing check that the strides must have the same direction for now (extra tests added in 5138ccd0e479d1702c2adefe1cd611e1aebe2f8b). Without this check, we would at least try to vectorize with runtime checks in cases where the runtime checks would always be false.

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp:2054]: Follow-up now available: https://github.com/llvm/llvm-project/pull/90036

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 10 above)
Comment (Meinersbur) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp:2123]: Not that I like the arbitrary-looking `!!CommonStride` scattered around the code, but at least I understand now. `FoundNonConstantDistanceDependence` sounded like something that's safer to set to true when in doubt. Should really be called `ReconsiderWithRuntimeChecks`. If I may suggest an alternative TODO:
```
// TODO: FoundNonConstantDistanceDependence is used as a necessary condition to consider retrying
// with runtime checks. Historically, we did not set it when strides were different but there is no inherent
// reason to.
```

**Review by Meinersbur — APPROVED** [review]
Thanks for explaining the code for me.

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 10 above)
Comment (fhahn) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp]: Thanks, adjusted as suggested and replace `!!CommonStride` with `CommonStride.has_value()`

**Comment by psteinfeld** [comment]
@fhahn, after this change, we started failing one of our tests -- the Fortran Polyhedron benchmark "rnflow".  When I build flang-new and compile the benchmark with the "flang-new -Ofast -march=native rnflow.f90", then run the benchmark, and then run the Polyhedron utility "pbvalid", you'll see failures.  The failures only happen when both compilation options are used.

Do you have access to the Polyhedron "rnflow" benchmark?

**Comment by fhahn** [comment]
@psteinfeld thanks for the heads up. Unfortunately I don't think I have access to that benchmark and I am also not really familiar with flang. Could you share the IR with and without the patch? It is likely that the patch enabled additional vectorization.

**Comment by vzakhari** [comment]
Regarding `rnflow` benchmark: yes, this change causes a loop vectorization with a memcheck.  The loop contains an FP division, which is being rewritten with Newton-Raphson division approximation with RCP in ISel. The approximation uses muls/adds/subs, and ISel encodes them as FMAs.  This altogether causes the accuracy change that affects the benchmark results.

Without this change, the loop is not vectorized and the scalar FP division is not selected into Newton-Raphson approximation. Gfortran does not vectorize the loop and also uses divss for the division.

It looks like the benchmark is sensitive to FP computations accuracy (that eventually affects even the integer results of the benchmark). Disabling FMA generation resolves the accuracy issue.

**Comment by psteinfeld** [comment]
> Newton-Raphson division approximation
Thanks, @vzakhari.  I'll try to "fix" this by disabling one of the options that affect lower level transformations.

@fhahn, there's nothing for you to do here.

**Comment by fhahn** [comment]
@vzakhari @psteinfeld good to know,  thanks for checking!
